scalar-crypto: Add zbkb, zbkc, zbkx extension listings.

ben-marshall opened the pull request:
As asked for by the scalar cryptography extension architecture reviewers, here are the instruction and extension definitions for instructions used by the scalar cryptography extension but that were originally defined by the Bit-manipulation Task Group.

Areas to double check for reviewers:

- Are the encodings correct, especially for `grevi`/`[un]shfli` derived instructions `rev.b`, `zip` and `unzip`.
- Do the instruction functional definitions match those in the _old_ LaTeX specification, or Spike?

I expect there will be iterations on this PR, hence my putting it in earlier than I originally planned, as I will be away most of next week.

---

*About the Sail code.* I struggled with this. I tried to make it pass the type checker, but quickly ran into the limits of my ability to write clear Sail which _also_ passes it's type checking. Hence, it hasn't been tested in an executable sort of way. For example, it is invalid to write:

    foreach (i from 0 to xlen) { <BODY> }

One must instead write

    foreach (i from 0 to sizeof(xlen)) { <BODY> }

And writing

    foreach (i from 0 to sizeof(xlen)-1) { <BODY> }

to avoid exceeding the bound of an array (i.e. 0..31 rather than 0..32) will cause a _syntax error_ on the "-1". The `foreach` construct is not documented completely in the Sail manual either it seems.

It's concatenation operator (`@`) also cannot (yet?) concatenate _single bits_ into a bit vector. But _can_ concatenate `N>1` length bit vectors together.

Where supporting code is needed to illustrate complete instruction functionality, this is included directly. I expect a broader scheme for managing "supporting Sail code" will be needed. The Cryptography extension particularly make many calls out to auxiliary code.

Generally, there is a tension between writing valid Sail code, and writing clearly comprehensible functional specifications. It is a great tool for lots of things, but sometimes, Sail and it's ruthless type-checking don't let you express things intuitively or obviously, with ugly results. I would be very keen to hear from Sail experts about better ways to express the instructions being added in this PR. See [the attached Sail file](https://github.com/riscv/riscv-bitmanip/files/6624629/riscv_insts_zbk.txt) I used to develop it.

topperc commented on bitmanip/insns/revb.adoc:
```diff
@@ -0,0 +1,54 @@
+===============================================================
+The bits-in-byte-reverse operation is only available for the full register
+width.  To emulate word-sized and halfword-sized reversal,
+perform a `rev.b rd,rs` followed by a `srai rd,rd`.
```
Is this note correct? If it's only reversing bits in each byte I don't think you need a shift for word or halfword. You can just ignore the upper bytes. Byte swap needs the shift because the lower bytes all end up in the upper part of the register. Reversing inside a byte doesn't change the position of the bytes.

aswaterman replied:
One wrinkle is that it doesn't preserve the sign-extension property for canonicalized 32-bit words in RV64 (specifically, if rs1[31] != rs1[24]).  So if that's required, it must be followed up with a `sext.w rd, rd`.

(Obviously, in some cases, that's unnecessary, e.g., if the result is stored to memory as a 32-bit word and the register becomes dead afterwards.)

ben-marshall replied:
Ah, yes, I must have forgotten to modify the hint after using another file as a template. I will remove it for now, I think the behaviour is fairly clear anyway.

topperc commented on bitmanip/insns/zip.adoc:
```diff
@@ -0,0 +1,73 @@
+....
+
+Description:: 
+This instruction gathers all of the odd and even bits of a source word into
```
I think the descriptions of zip and unzip are reversed. I thought zip interleaves the bits into even and odd bits like a zipper.

ben-marshall replied:
Right you are!

topperc commented on bitmanip/insns/packw.adoc:
```diff
@@ -0,0 +1,53 @@
+Description:: 
+This instruction packs the low 16 bits of
+_rs1_ and _rs2_ into the 32 least-significant bits of _rd_,
+zero extending the rest of _rd_.
```
It was not well spelled out in the old spec, but I would have expected packw to sign extend the results like other w instructions.

aswaterman replied:
Yes, it should be sign extension.  The old spec actually did say this, just not in the most obvious way.  At the top of Chapter 2, it says: "Like in the base ISA, we add *W instruction variants on RV64 with the semantic of the matching RV32 instruction. Those instructions ignore the upper 32 bit of their input and sign-extend their 32 bit output values."

aswaterman replied:
(Also, on bitmanip matters, I recommend trusting Spike when there's any perceived ambiguity; it is quite well vetted and was written by the same person who wrote the original spec.  Spike does sign extend.)

ben-marshall replied:
Okay, I'll refer to Spike as the canonical example then. I totally missed the sign extend bit in the spec, I'll add that in.

aswaterman commented on bitmanip/insns/pack.adoc:
```diff
@@ -0,0 +1,56 @@
+    let lo_half : bits(32) = X(rs1)[31..0];
+    let hi_half : bits(32) = X(rs2)[31..0];
+    X(rd) = EXTZ(hi_half @ lo_half);
+};
```
This is just aesthetic, but can we compact this as

    let lo_half : bits(xlen/2) = X(rs1)[xlen/2-1..0];
    let hi_half : bits(xlen/2) = X(rs2)[xlen/2-1..0];
    X(rd) = EXTZ(hi_half @ lo_half);

I find this version easier to digest, because I don't have to think hard about how the RV32 and RV64 variants differ.

ben-marshall replied:
So. This is a good example of _do you want a clear spec or do you want valid Sail?_ I'm in favour of your example, so I'll use that for now, but this will keep coming up I think.

aswaterman commented on bitmanip/insns/packw.adoc:
```diff
@@ -0,0 +1,53 @@
+--
+let lo_half : bits(16) = X(rs1)[15..0];
+let hi_half : bits(16) = X(rs2)[15..0];
+X(rd) = EXTZ(hi_half @ lo_half);
```
EXTS

aswaterman commented on bitmanip/insns/pack.adoc:
```diff
@@ -0,0 +1,56 @@
+Operation::
+[source,sail]
+--
+if (sizeof(xlen) == 32) then {
```
`xlen` not `sizeof(xlen)` right?

aswaterman commented on bitmanip/insns/revb.adoc:
```diff
@@ -0,0 +1,54 @@
+[source,sail]
+--
+result : xlenbits = EXTZ(0b0);
+foreach (i from 0 to sizeof(xlen) by 8) {
```
`xlen` not `sizeof(xlen)` right?

aswaterman commented on bitmanip/insns/unzip.adoc:
```diff
@@ -0,0 +1,73 @@
+    x = shfl_stage(x, 0x00ff0000, 0x0000ff00, 0x8);
+    X(rd) = x;
+    RETIRE_SUCCESS
+}
```
I think we should not express this computation in terms of the general shuffle operations, but instead just say what it does directly.  I'm thinking something like this (pardon likely incorrect syntax):

    foreach (i from 0 to xlen/2-1) {
      X(rd)[i] = X(rs1)[2*i]
      X(rd)[i+xlen/2] = X(rs1)[2*i+1]
    }

ben-marshall replied:
Again, I agree this is much clearer than the Sail I've written, I'll use this.

aswaterman commented on bitmanip/insns/zip.adoc:
```diff
@@ -0,0 +1,73 @@
+    x = shfl_stage(x, 0x44444444, 0x22222222, 0x1);
+    X(rd) = x;
+    RETIRE_SUCCESS
+}
```
I think we should not express this computation in terms of the general shuffle operations, but instead just say what it does directly.  I'm thinking something like this (pardon likely incorrect syntax):

    foreach (i from 0 to xlen/2-1) {
      X(rd)[2*i] = X(rs1)[i]
      X(rd)[2*i+1] = X(rs1)[i+xlen/2]
    }

aswaterman commented on bitmanip/insns/xpermn.adoc:
```diff
@@ -0,0 +1,60 @@
+--
+val xpermn_lookup : (bits(4), xlenbits) -> bits(4)
+function xpermn_lookup (idx, lut) = {
+    (lut >> (idx @ 0b00))[3..0]
```
Is this concatenation backwards (i.e. should it not be `0b00 @ idx`, or even just `idx`)?

ben-marshall replied:
It looks wrong, but the index is in the range 0..N _nibbles_, and must be multiplied by `4` to get a _bitwise_ shift amount.

ben-marshall replied:
I'll be honest, these were by far the hardest instructions to express "nicely" in _valid_ Sail.

aswaterman commented on bitmanip/insns/xpermb.adoc:
```diff
@@ -0,0 +1,60 @@
+--
+val xpermb_lookup : (bits(8), xlenbits) -> bits(8)
+function xpermb_lookup (idx, lut) = {
+    (lut >> (idx @ 0b000))[7..0]
```
Is this concatenation backwards (i.e. should it not be 0b000 @ idx, or even just idx)?

ben-marshall replied:
Likewise the index is in the range 0..N _bytes_, and must be multiplied by `8` to get a _bitwise_ shift amount.

ben-marshall commented:
Okay, I've tried to address everything in @aswaterman's and @topperc's comments. Repeated from the commit message:

- Changed `pack` instruciton description to be clearer. No longer valid Sail.
- Fixed `packw` -> changed zero extend to sign extend.
- Removed incorrect software hint for `revb`.
- Flipped "gathers"/"scatters" for `zip` and `unzip`, as I had them the wrong way around.
- Changed expression of `zip` and `unzip` to be invalid Sail but _much_ clearer based on Andrew's suggestions.

ben-marshall commented:
I've just updated this PR with feedback from the scalar crypto architecture review. This will make it's way into v0.9.4 of the scalar crypto specification.

aswaterman commented:
LGTM now.  Thank you, @ben-marshall!

lhtin commented on bitmanip/zbkx.adoc:
```diff
@@ -0,0 +1,38 @@
+inside the general purpose registers.
+_rs1_ is used as a vector of N-bit words, and _rs2_ as a vector of N-bit
+indicies into _rs1_.
+Elements in _rs1_ are replaced by the indexed element in _rs2_, or zero
```
May I ask if the `rs1` here should be changed to `rd`? Shouldn't we be modifying the contents of `rd` instead of `rs1`?

lhtin replied:
Also shouldn't `rs2` be changed to `rs1` here? Because earlier it says `rs1` is a data array and `rs2` is an index array.

aswaterman replied:
> May I ask if the rs1 here should be changed to rd? Shouldn't we be modifying the contents of rd instead of rs1?

I understand what the spec is trying to say here: if you append ", and then written to rd" to the end of the sentence, it becomes clearer.  In any case, the SAIL code makes it unambiguous.

> Also shouldn't rs2 be changed to rs1 here? Because earlier it says rs1 is a data array and rs2 is an index array.

It would be clearer to say "element named by rs2" rather than "element in rs2", but the preceding sentence and the SAIL code make it unambiguous.

lhtin replied:
The SAIL code is really clear. I've looked at the description of the corresponding xperm.n directive and feel that the two descriptions are a bit ambiguous (one says `elements in rs1 are replaced...`, the other says `each element in rs2 replaced...`):

> The xperm.n instruction operates on nibbles. The rs1 register contains a vector of XLEN/4 4-bit elements. The rs2 register contains a vector of XLEN/4 4-bit indexes. The result is each element in rs2 replaced by the indexed element in rs1, or zero if the index into rs2 is out of bounds.
